trunk/source/Library-CHC/src/de/uni_freiburg/informatik/ultimate/lib/chc/SmtChcScript.java

```diff
@@ -0,0 +1,157 @@
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with the ULTIMATE CHC Library. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Additional permission under GNU GPL version 3 section 7:
+ * If you modify the ULTIMATE CHC Library, or any covered work, by linking
+ * or combining it with Eclipse RCP (or a modified version of Eclipse RCP),
+ * containing parts covered by the terms of the Eclipse Public License, the
+ * licensors of the ULTIMATE CHC Library grant you additional permission
+ * to convey the resulting work.
+ */
+package de.uni_freiburg.informatik.ultimate.lib.chc;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import de.uni_freiburg.informatik.ultimate.lib.smtlibutils.ManagedScript;
+import de.uni_freiburg.informatik.ultimate.logic.ApplicationTerm;
+import de.uni_freiburg.informatik.ultimate.logic.Model;
+import de.uni_freiburg.informatik.ultimate.logic.SMTLIBConstants;
+import de.uni_freiburg.informatik.ultimate.logic.Script;
+import de.uni_freiburg.informatik.ultimate.logic.Script.LBool;
+
+/**
+ * Used to access a constraint Horn solver via the {@link Script} interface. This can e.g. be used to access Z3's CHC
+ * engine, Spacer.
+ *
+ * @author Dominik Klumpp (klumpp@informatik.uni-freiburg.de)
+ */
+public class SmtChcScript implements IChcScript {
+	private static final boolean ADD_COMMENTS = false;
+
+	// We assume the function symbols are already declared in the script, as the Horn clause terms use the same script.
+	// To support different scripts for creating and solving Horn clauses, we would have to transfer terms and declare
+	// the necessary function symbols.
+	private static final boolean DECLARE_FUNCTIONS = false;
+
+	private final ManagedScript mMgdScript;
+	private boolean mProduceUnsatCores;
+	private Map<String, HornClause> mName2Clause;
+
+	public SmtChcScript(final ManagedScript mgdScript) {
+		mMgdScript = mgdScript;
+		mMgdScript.lock(this);
+	}
+
+	@Override
+	public Script getScript() {
+		return mMgdScript.getScript();
+	}
+
+	@Override
+	public LBool solve(final HcSymbolTable symbolTable, final List<HornClause> system) {
+		reset();
+
+		mMgdScript.unlock(this);
+		final var asserter =
+				new ChcAsserter(mMgdScript, getScript(), mProduceUnsatCores, ADD_COMMENTS, DECLARE_FUNCTIONS);
+		asserter.assertClauses(symbolTable, system);
+		mMgdScript.lock(this);
+
+		mName2Clause = asserter.getName2Clause();
+
+		return mMgdScript.checkSat(this);
+	}
+
+	@Override
+	public LBool solve(final HcSymbolTable symbolTable, final List<HornClause> system, final long timeout) {
+		mMgdScript.getScript().setOption(":timeout", timeout);
+		try {
+			return solve(symbolTable, system);
+		} finally {
+			try {
+				// reset timeout
+				mMgdScript.getScript().setOption(":timeout", 0L);
```

Comment:
this looks _really_ hacky

Reply:
Yes. I don't know if there is a better solution, do you have an idea?